pkg/pstore/api.go

```diff
@@ -0,0 +1,149 @@
+// Copyright (C) 2018 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package pstore // import "storj.io/storj/pkg/pstore"
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"os"
+	"path"
+)
+
+type argError struct {
+	arg string
+	msg string
+}
+
+func (e *argError) Error() string {
+	return fmt.Sprintf("HashError (%s): %s", string(e.arg), e.msg)
+}
+
+type fsError struct {
+	path string
+	msg  string
+}
+
+func (e *fsError) Error() string {
+	return fmt.Sprintf("FSError (%s): %s", e.path, e.msg)
+}
+
+func Store(hash string, r *bufio.Reader, dir string) error {
+	fmt.Println("Storing...")
+	if len(hash) < 20 {
+		return &argError{hash, "Hash is too short"}
+	}
+	if dir == "" {
+		return &argError{dir, "No path provided"}
+	}
+
+	// Folder structure
+	folder1 := string(hash[0:2])
+	folder2 := string(hash[2:4])
+	fileName := string(hash[4:])
+
+	// Create directory path string
+	dirpath := path.Join(dir, folder1, folder2)
+
+	// Create directory path on file system
+	mkDirerr := os.MkdirAll(dirpath, 0700)
+	if mkDirerr != nil {
+		return mkDirerr
+	}
+
+	// Create File Path string
+	filepath := path.Join(dirpath, fileName)
+
+	// Create File on file system
+	file, openErr := os.OpenFile(filepath, os.O_RDWR|os.O_CREATE, 0755)
+	if openErr != nil {
+		return openErr
+	}
+
+	// Close when finished
+	defer file.Close()
+
+	// Buffer for reading data
+	buffer := make([]byte, 4096)
+	for {
+		// Read data from read stream into buffer
+		n, err := r.Read(buffer)
+		if err == io.EOF {
+			break
+		}
+
+		// Write the buffer to the file we opened earlier
+		_, err = file.Write(buffer[:n])
+	}
+
+	return nil
+}
+
+func Retrieve(hash string, w *bufio.Writer, dir string) error {
```

Comment:
it'd be awesome if you added function comments like 

```
// Retrieve will get a file identified by hash and send it to the given writer
func Retrieve(...
```

though, that said, I think in an earlier review I suggested doing something like
```
func Retrieve(hash string, dir string) (ranger.Ranger, error)
```

instead. You can turn a filehandle into a ranger.Ranger by using ranger.ReaderAtRanger